// tock-node/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function i18nText(text) {
    return {
        text: text,
        args: [],
        toBeTranslated: true,
        length: text.length,
    };
}
exports.i18nText = i18nText;
function suggestion(label) {
    return {
        title: i18nText(label),
    };
}
exports.suggestion = suggestion;
function action(label, url) {
    return {
        title: i18nText(label),
        url: url,
    };
}
exports.action = action;
function imageCard(title, img, subTitle, quickReplies) {
    if (subTitle === void 0) { subTitle = ''; }
    if (quickReplies === void 0) { quickReplies = []; }
    return {
        actions: quickReplies,
        attachment: {
            url: img,
            type: 'image',
        },
        delay: 0,
        subTitle: i18nText(subTitle),
        title: i18nText(title),
        type: 'card',
    };
}
exports.imageCard = imageCard;
function videoCard(title, video, subTitle, quickReplies) {
    if (subTitle === void 0) { subTitle = ''; }
    if (quickReplies === void 0) { quickReplies = []; }
    return {
        actions: quickReplies,
        attachment: {
            url: video,
            type: 'video',
        },
        delay: 0,
        subTitle: i18nText(subTitle),
        title: i18nText(title),
        type: 'card',
    };
}
exports.videoCard = videoCard;
function audioCard(title, audio, subTitle, quickReplies) {
    if (subTitle === void 0) { subTitle = ''; }
    if (quickReplies === void 0) { quickReplies = []; }
    return {
        actions: quickReplies,
        attachment: {
            url: audio,
            type: 'audio',
        },
        delay: 0,
        subTitle: i18nText(subTitle),
        title: i18nText(title),
        type: 'card',
    };
}
exports.audioCard = audioCard;
function fileCard(title, file, subTitle, quickReplies) {
    if (subTitle === void 0) { subTitle = ''; }
    if (quickReplies === void 0) { quickReplies = []; }
    return {
        actions: quickReplies,
        attachment: {
            url: file,
            type: 'file',
        },
        delay: 0,
        subTitle: i18nText(subTitle),
        title: i18nText(title),
        type: 'card',
    };
}
exports.fileCard = fileCard;
